sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeCreator.scala

```diff
@@ -174,59 +174,32 @@ case class CreateMap(children: Seq[Expression]) extends Expression {
 /**
  * Returns a Row containing the evaluation of all children expressions.
  */
-@ExpressionDescription(
-  usage = "_FUNC_(col1, col2, col3, ...) - Creates a struct with the given field values.")
-case class CreateStruct(children: Seq[Expression]) extends Expression {
-
-  override def foldable: Boolean = children.forall(_.foldable)
-
-  override lazy val dataType: StructType = {
-    val fields = children.zipWithIndex.map { case (child, idx) =>
-      child match {
-        case ne: NamedExpression =>
-          StructField(ne.name, ne.dataType, ne.nullable, ne.metadata)
-        case _ =>
-          StructField(s"col${idx + 1}", child.dataType, child.nullable, Metadata.empty)
+object CreateStruct extends (Seq[Expression] => CreateNamedStruct) {
+  def apply(children: Seq[Expression]) : CreateNamedStruct = {
+    val namedStructArgs = {
+      val attNames = for {
+        (child, idx) <- children.zipWithIndex
+      } yield {
+        child match {
+          case ne: NamedExpression => ne.name
+          case _ => s"col${idx + 1}"
```

Comment:
I'd argue that we break it already. In the previous `CreateStruct`, the field name is inherited from **resolved** `NamedExpression`, but now we get it from the `NamedExpression` that may not be resolved. This can bring inconsistency, e.g. after resolve `UnresolvedAttribute` to `AttriuteReference`, the name may change.

If you agree that default column name is not a backwards compatibility issue, why do you treat default struct field name a backwards compatibility issue?